From 09505a04a50d70be6f51e5ca43e6e11532509726 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 27 Apr 2022 22:59:32 +0800
Subject: [PATCH 06/90] Revert "gl-renderer: require GL_EXT_unpack_subimage"

This reverts commit 593d5af43a8e2c2a3371088fa7ae430d0517c82d.

For old mali, e.g. utgard.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/renderer-gl/gl-renderer-internal.h |  2 ++
 libweston/renderer-gl/gl-renderer.c          | 32 +++++++++++++++++---
 2 files changed, 29 insertions(+), 5 deletions(-)

diff --git a/libweston/renderer-gl/gl-renderer-internal.h b/libweston/renderer-gl/gl-renderer-internal.h
index 888df847..cabf1ed0 100644
--- a/libweston/renderer-gl/gl-renderer-internal.h
+++ b/libweston/renderer-gl/gl-renderer-internal.h
@@ -153,6 +153,8 @@ struct gl_renderer {
 	PFNEGLCREATEPLATFORMWINDOWSURFACEEXTPROC create_platform_window;
 	bool has_platform_base;
 
+	bool has_unpack_subimage;
+
 	PFNEGLBINDWAYLANDDISPLAYWL bind_display;
 	PFNEGLUNBINDWAYLANDDISPLAYWL unbind_display;
 	PFNEGLQUERYWAYLANDBUFFERWL query_buffer;
diff --git a/libweston/renderer-gl/gl-renderer.c b/libweston/renderer-gl/gl-renderer.c
index 7f5d35d5..5a0cd4c7 100644
--- a/libweston/renderer-gl/gl-renderer.c
+++ b/libweston/renderer-gl/gl-renderer.c
@@ -1970,6 +1970,7 @@ gl_renderer_flush_damage(struct weston_surface *surface,
 {
 	const struct weston_testsuite_quirks *quirks =
 		&surface->compositor->test_data.test_quirks;
+	struct gl_renderer *gr = get_renderer(surface->compositor);
 	struct gl_surface_state *gs = get_surface_state(surface);
 	struct gl_buffer_state *gb = gs->buffer;
 	struct weston_view *view;
@@ -2012,6 +2013,27 @@ gl_renderer_flush_damage(struct weston_surface *surface,
 
 	glActiveTexture(GL_TEXTURE0);
 
+	if (!gr->has_unpack_subimage) {
+		wl_shm_buffer_begin_access(buffer->shm_buffer);
+		for (j = 0; j < gb->num_textures; j++) {
+			int hsub = pixel_format_hsub(buffer->pixel_format, j);
+			int vsub = pixel_format_vsub(buffer->pixel_format, j);
+
+			glBindTexture(GL_TEXTURE_2D, gb->textures[j]);
+			glTexImage2D(GL_TEXTURE_2D, 0,
+				     gb->gl_format[j],
+				     gb->pitch / hsub,
+				     buffer->height / vsub,
+				     0,
+				     gl_format_from_internal(gb->gl_format[j]),
+				     gb->gl_pixel_type,
+				     data + gb->offset[j]);
+		}
+		wl_shm_buffer_end_access(buffer->shm_buffer);
+
+		goto done;
+	}
+
 	if (gb->needs_full_upload || quirks->gl_force_full_upload) {
 		glPixelStorei(GL_UNPACK_SKIP_PIXELS_EXT, 0);
 		glPixelStorei(GL_UNPACK_SKIP_ROWS_EXT, 0);
@@ -4103,11 +4125,9 @@ gl_renderer_setup(struct weston_compositor *ec, EGLSurface egl_surface)
 	else
 		ec->read_format = pixel_format_get_info_by_pixman(PIXMAN_a8b8g8r8);
 
-	if (gr->gl_version < gr_gl_version(3, 0) &&
-	    !weston_check_egl_extension(extensions, "GL_EXT_unpack_subimage")) {
-		weston_log("GL_EXT_unpack_subimage not available.\n");
-		return -1;
-	}
+	if (gr->gl_version >= gr_gl_version(3, 0) ||
+	    weston_check_egl_extension(extensions, "GL_EXT_unpack_subimage"))
+		gr->has_unpack_subimage = true;
 
 	if (gr->gl_version >= gr_gl_version(3, 0) ||
 	    weston_check_egl_extension(extensions, "GL_EXT_texture_type_2_10_10_10_REV"))
@@ -4206,6 +4226,8 @@ gl_renderer_setup(struct weston_compositor *ec, EGLSurface egl_surface)
 			    yesno(gr->has_gl_texture_rg));
 	weston_log_continue(STAMP_SPACE "OES_EGL_image_external: %s\n",
 			    yesno(gr->has_egl_image_external));
+	weston_log_continue(STAMP_SPACE "wl_shm sub-image to texture: %s\n",
+			    gr->has_unpack_subimage ? "yes" : "no");
 
 	return 0;
 }
-- 
2.20.1

